akka-http-core/src/main/scala/akka/http/engine/client/HostConnectionPoolGateway.scala

```diff
@@ -0,0 +1,139 @@
+/**
+ * Copyright (C) 2009-2014 Typesafe Inc. <http://www.typesafe.com>
+ */
+
+package akka.http.engine.client
+
+import scala.annotation.tailrec
+import scala.concurrent.{ Future, Promise }
+import scala.concurrent.duration.{ Duration, FiniteDuration }
+import scala.util.{ Failure, Try }
+import akka.actor.{ ActorLogging, Cancellable }
+import akka.stream.FlowMaterializer
+import akka.stream.actor.{ ActorPublisher, ActorSubscriber, ZeroRequestStrategy }
+import akka.stream.actor.ActorPublisherMessage._
+import akka.stream.actor.ActorSubscriberMessage._
+import akka.stream.impl.ResizableMultiReaderRingBuffer
+import akka.stream.scaladsl.{ Sink, Source, Flow }
+import akka.http.model._
+
+private object HostConnectionPoolGateway {
+  import ConnectionPool.{ PoolRequest, BeginShutdown }
+
+  case class RequestContext(request: HttpRequest, responsePromise: Promise[HttpResponse], retriesLeft: Int) {
+    require(retriesLeft >= 0)
+  }
+  case class ResponseContext(rc: RequestContext, response: Try[HttpResponse])
+
+  private case object ShutdownNow
+
```

Comment:
Please add a comment that this Actor wraps a Flow, so its Publisher and Subscriber interfaces are “internal”.